web/handlers/genomics/util.py

```diff
@@ -209,48 +209,28 @@ def get_major_lineage_prevalence(df, index_col = "date", min_date = None, max_da
    
     df['prevalence'] = df['total_count']/df['lineage_count']
     df = df.sort_values(by="date") #Sort date values
-    min_date = dt.strptime(min_date, "%Y-%m-%d")
-    max_date = dt.strptime(max_date, "%Y-%m-%d")
+    
     
     if min_date and max_date:
         df = df[(df["date"].between(min_date, max_date)) & (df["prevalence"] >= prevalence_threshold)]
-              
+        num_unique_dates = df[df["date"] >= min_date]["date"].unique().shape[0] #counts # of unique days lineage is found
     elif min_date:
-        date_limit = dt.strptime(min_date, "%Y-%m-%d") + timedelta(days=ndays)
+        date_limit = dt.strptime(min_date, "%Y-%m-%d") + timedelta(days=ndays) # searches from min_date to ndays forward
         df = df[(df["prevalence"] >= prevalence_threshold) & (df['date'] > min_date) & (df['date'] < date_limit)]
-        
-    elif max_date:
-        date_limit = dt.strptime(max_date, "%Y-%m-%d") - timedelta(days=ndays)
-        df = df[(df["prevalence"] >= prevalence_threshold) & (df['date'] < max_date) & (df['date'] > date_limit)]
-    
-        
-    if df["date"].iloc[-1] < dt.today():
-            date = df["date"].iloc[-1]
+        num_unique_dates = df[df["date"] <= date_limit]["date"].unique().shape[0] 
     else:
-            date = dt.today()
-            date_range = date - timedelta(days = ndays)
-            df = df[(df["prevalence"] >= prevalence_threshold) & (df['date'] > date_range) & (df['date'] < date)]
-    
-    date_limit = date - timedelta(days = ndays)
-    num_unique_dates = df[df["date"] >= date_limit]["date"].unique().shape[0]
-    if num_unique_dates < nday_threshold: 
-         nday_threshold = round((nday_threshold/ndays) * num_unique_dates)
-         
-    date_range = date - timedelta(days = nday_threshold) #Finding lineages nday_threshold days in the date range
-    lineages_to_retain = df[(df["prevalence"] >= prevalence_threshold) & (df['date'] > date_range) & (df['date'] < date)]
-    lineages_to_retain = lineages_to_retain['lineage'].to_list()
-    keep_lineages.append(lineages_to_retain)
+        date_limit = dt.strptime(max_date, "%Y-%m-%d") - timedelta(days=ndays) # searches from max_date to ndays back
+        df = df[(df["prevalence"] >= prevalence_threshold) & (df['date'] < max_date) & (df['date'] > date_limit)]
+        num_unique_dates = df[df["date"] >= date_limit]["date"].unique().shape[0]
     
-    df = df.groupby(index_col).apply(classify_other_category, keep_lineages)
-
+    if num_unique_dates < nday_threshold:
+        nday_threshold = round((nday_threshold/ndays) * num_unique_dates) 
+    lineage_counts = df["lineage"].value_counts() #number of times lineage is found in df
```

Comment:
If you put your prevalence threshold filter down here instead, you should get the same counts as your current version, but there won't be gaps in the dataframe on low prevalence days